x-pack/plugins/apm/server/lib/search_strategies/correlations/async_search_service.ts

```diff
@@ -0,0 +1,264 @@
+import { fetchTransactionDurationPecentiles } from './query_percentiles';
+import { fetchTransactionDurationCorrelation } from './query_correlation';
+import { fetchTransactionDurationHistogramRangesteps } from './query_histogram_rangesteps';
+import { fetchTransactionDurationRanges, HistogramItem } from './query_ranges';
+import { hashHistogram, isHistogramRoughlyEqual } from './utils';
+import { asPreciseDecimal } from '../../../../common/utils/formatters';
+import type {
+  AsyncSearchProviderProgress,
+  SearchServiceParams,
+  SearchServiceValue,
+} from '../../../../common/search_strategies/correlations/types';
+
+const CORRELATION_THRESHOLD = 0.3;
+const KS_TEST_THRESHOLD = 0.1;
+const SIGNIFICANT_FRACTION = 3;
+
+interface HashedSearchServiceValue {
+  histogram: HistogramItem[];
+  value: string;
+  field: string;
+  correlation: number;
+  ksTest: number;
+  duplicatedFields: Set<string>;
+}
+export const asyncSearchServiceProvider = (
+  esClient: ElasticsearchClient,
+  params: SearchServiceParams
+) => {
+  let isCancelled = false;
+  let isRunning = true;
+  let error: Error;
+
+  const progress: AsyncSearchProviderProgress = {
+    started: Date.now(),
+    loadedHistogramStepsize: 0,
+    loadedOverallHistogram: 0,
+    loadedFieldCanditates: 0,
+    loadedFieldValuePairs: 0,
+    loadedHistograms: 0,
+    getOverallProgress: () =>
+      progress.loadedHistogramStepsize * 0.025 +
+      progress.loadedOverallHistogram * 0.025 +
+      progress.loadedFieldCanditates * 0.025 +
+      progress.loadedFieldValuePairs * 0.025 +
+      progress.loadedHistograms * 0.9,
+  };
+
+  const values: SearchServiceValue[] = [];
+  let overallHistogram: HistogramItem[] | undefined;
+
+  let percentileThresholdValue: number;
+
+  const cancel = () => {
+    isCancelled = true;
+  };
+
+  const fetchCorrelations = async () => {
+    try {
+      // 95th percentile to be displayed as a marker in the log log chart
+      const percentileThreshold = await fetchTransactionDurationPecentiles(
+        esClient,
+        params,
+        params.percentileThreshold ? [params.percentileThreshold] : undefined
+      );
+      percentileThresholdValue =
+        percentileThreshold[`${params.percentileThreshold}.0`];
+
+      const histogramRangeSteps = await fetchTransactionDurationHistogramRangesteps(
+        esClient,
+        params
+      );
+      progress.loadedHistogramStepsize = 1;
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const overallLogHistogramChartData = await fetchTransactionDurationRanges(
+        esClient,
+        params,
+        histogramRangeSteps
+      );
+      progress.loadedOverallHistogram = 1;
+      overallHistogram = overallLogHistogramChartData;
+
+      if (isCancelled) {
+        isRunning = false;
+        return;
+      }
+
+      const {
+        fieldCandidates,
+        totalHits,
+      } = await fetchTransactionDurationFieldCandidates(esClient, params);
```

Comment:
`totalHits` here cannot be used when we do these calculations against `histogram` fields. 

`histogram` fields are pre-aggregated fields, so, that means the doc hit count does not accurately match the actual data count.

What should happen instead is:

- Calculate the true fractions given the deduplicated percentile ranges (see other comment about the fractions calculation)
- Use the sum of the values as your `totalHit`

Reply:
As discussed, I will address this in the next follow up.